### PR TITLE
SDL_Test: format tracked allocation names during report generation

### PR DESCRIPTION
**Pull request by madebr**
On my Linux system, `testsprite --trackmem` goes from  64 to 3200 fps.

### CODE REVIEW

**Comment by slouken** [comment]
With this change, does it make sense to enable `--trackmem` for tests in CI?

**Comment by madebr** [comment]
> With this change, does it make sense to enable `--trackmem` for tests in CI? And possibly even failing CI for leaked memory?

It might, but ci/ctest only tests dummy audio/video.
So my latest leak reports on linux/windows would not trigger a ci failure.

**Comment by slouken** [comment]
That's fine, but it seems like a worthwhile change to make, to catch new leaks or regressions.

**Comment by madebr** [comment]
I added a cmake option to add `--trackmem` when supported (it's opt-out).

Apps using SDL main callbacks cannot fail due to outstanding memory allocations.
`SDL_AppQuit` does not allow to override the return code:
https://github.com/libsdl-org/SDL/blob/35e42d0a254f230bb736cf993a9c3855ea1680a1/include/SDL3/SDL_main.h#L199


On my system, `testthread --trackmem` fails with the following leaks:
```
INFO: Memory allocations:
Allocation 0: 4 bytes
	0x7fbdd7117f28: SDL_malloc_REAL+0x31
	0x7fbdd711bbed: SDL_strdup_REAL+0x32
	0x7fbdd711e22f: SDL_CreateThreadWithPropertiesRuntime_REAL+0x13a
	0x7fbdd711e34a: SDL_CreateThreadRuntime_REAL+0x87
	0x7fbdd7064933: SDL_CreateThreadRuntime+0x46
	0x40506c: main+0x266
	0x7fbdd6e4614a: __libc_start_call_main+0x7a
	0x7fbdd6e4620b: __libc_start_main+0x8b
	0x404b45: _start+0x25
Allocation 1: 112 bytes
	0x7fbdd7117f9f: SDL_calloc_REAL+0x4b
	0x7fbdd711e1e7: SDL_CreateThreadWithPropertiesRuntime_REAL+0xf2
	0x7fbdd711e34a: SDL_CreateThreadRuntime_REAL+0x87
	0x7fbdd7064933: SDL_CreateThreadRuntime+0x46
	0x40506c: main+0x266
	0x7fbdd6e4614a: __libc_start_call_main+0x7a
	0x7fbdd6e4620b: __libc_start_main+0x8b
	0x404b45: _start+0x25
Total: 0.11 Kb in 2 allocations
```

**Comment by madebr** [comment]
This is the backtrace of the segfault:
```
#0  0x000000000040eed9 in SDLTest_LogAllocations () at /home/maarten/projects/SDL/src/test/SDL_test_memory.c:342
#1  0x000000000040542b in SDLTest_CommonDestroyState (state=0x4b78c20)
    at /home/maarten/projects/SDL/src/test/SDL_test_common.c:156
#2  0x0000000000404c6a in quit (rc=0) at /home/maarten/projects/SDL/test/testthread.c:32
#3  0x0000000000404e03 in killed (sig=15) at /home/maarten/projects/SDL/test/testthread.c:86
#4  <signal handler called>
#5  0x00007f86dbaae834 in __pthread_kill_implementation () from /lib64/libc.so.6
#6  0x00007f86dba5c8ee in raise () from /lib64/libc.so.6
#7  0x00000000004050a7 in main (argc=2, argv=0x7fff7ea66878) at /home/maarten/projects/SDL/test/testthread.c:158
```

**Comment by madebr** [comment]
@icculus 
`SDL_AppQuit` functions cannot influence the exit code. Would it make sense to do so?
I'd like to return a non-zero value when there are outstanding memory allocations.

**Comment by icculus** [comment]
There will always be outstanding allocations there because SDL_Quit isn't called until it returns.

**Comment by madebr** [comment]
> There will always be outstanding allocations there because SDL_Quit isn't called until it returns.

Indeed, though, the tests call `SDL_Quit` in `SDL_AppQuit`.

**Comment by icculus** [comment]
Could we write something to stdout or touch a file to signify outstanding allocations? Or do the SDL_Quit+allocation checks at the point in SDL_AppIterate where we decide to end the program?

I don't want to change the API or add a hint if we can find another way to transmit this information.

**Comment by madebr** [comment]
> Could we write something to stdout or touch a file to signify outstanding allocations? Or do the SDL_Quit+allocation checks at the point in SDL_AppIterate where we decide to end the program?

`SDLTest_LogAllocations` logs to stderr (using `SDL_Log`).
I removed the need for a non-zero return code, and uses [a regular expression](https://cmake.org/cmake/help/latest/prop_test/FAIL_REGULAR_EXPRESSION.html) instead to detect leaks.

**Comment by madebr** [comment]
I can reproduce the `torturethread` issues on Linux.
This is a stacktrace:
```
#0  0x000000000040e72e in SDL_GetTrackedAllocation (mem=0x7ff6bc001cb0)
    at /home/maarten/projects/SDL/src/test/SDL_test_memory.c:85
#1  0x000000000040e7b2 in SDL_IsAllocationTracked (mem=0x7ff6bc001cb0)
    at /home/maarten/projects/SDL/src/test/SDL_test_memory.c:101
#2  0x000000000040e7fc in SDL_TrackAllocation (mem=0x7ff6bc001cb0, size=40)
    at /home/maarten/projects/SDL/src/test/SDL_test_memory.c:109
#3  0x000000000040eac6 in SDLTest_TrackedCalloc (nmemb=1, size=40)
    at /home/maarten/projects/SDL/src/test/SDL_test_memory.c:214
#4  0x00007ff704b183ee in SDL_calloc_REAL (nmemb=1, size=40)
    at /home/maarten/projects/SDL/src/stdlib/SDL_malloc.c:5306
#5  0x00007ff704cde9ac in SDL_CreateMutex_REAL ()
    at /home/maarten/projects/SDL/src/thread/pthread/SDL_sysmutex.c:34
#6  0x00007ff704a353e4 in SDL_CreateProperties_REAL () at /home/maarten/projects/SDL/src/SDL_properties.c:159
#7  0x00007ff704b1e739 in SDL_CreateThreadRuntime_REAL (fn=0x404c2e <SubThreadFunc>, name=0x7ff6c5fffd90 "Child8_5", 
    userdata=0x7ff6c5fffde4, pfnBeginThread=0x0, pfnEndThread=0x0)
    at /home/maarten/projects/SDL/src/thread/SDL_thread.c:399
#8  0x00007ff704a64ccd in SDL_CreateThreadRuntime (a=0x404c2e <SubThreadFunc>, b=0x7ff6c5fffd90 "Child8_5", 
    c=0x7ff6c5fffde4, d=0x0, e=0x0) at /home/maarten/projects/SDL/src/dynapi/SDL_dynapi_procs.h:127
#9  0x0000000000404d06 in ThreadFunc (data=0x8) at /home/maarten/projects/SDL/test/torturethread.c:60
#10 0x00007ff704b1e4de in SDL_RunThread (thread=0x13050740)
    at /home/maarten/projects/SDL/src/thread/SDL_thread.c:323
#11 0x00007ff704cde550 in RunThread (data=0x13050740)
    at /home/maarten/projects/SDL/src/thread/pthread/SDL_systhread.c:69
#12 0x00007ff7048ac897 in start_thread () from /lib64/libc.so.6
#13 0x00007ff704933a5c in clone3 () from /lib64/libc.so.6
```

**Comment by madebr** [comment]
> I can reproduce the `torturethread` issue on Linux. This is a stacktrace:
> 
> ```
> #0  0x000000000040e72e in SDL_GetTrackedAllocation (mem=0x7ff6bc001cb0)
>     at /home/maarten/projects/SDL/src/test/SDL_test_memory.c:85
> #1  0x000000000040e7b2 in SDL_IsAllocationTracked (mem=0x7ff6bc001cb0)
>     at /home/maarten/projects/SDL/src/test/SDL_test_memory.c:101
> #2  0x000000000040e7fc in SDL_TrackAllocation (mem=0x7ff6bc001cb0, size=40)
>     at /home/maarten/projects/SDL/src/test/SDL_test_memory.c:109
> #3  0x000000000040eac6 in SDLTest_TrackedCalloc (nmemb=1, size=40)
>     at /home/maarten/projects/SDL/src/test/SDL_test_memory.c:214
> #4  0x00007ff704b183ee in SDL_calloc_REAL (nmemb=1, size=40)
>     at /home/maarten/projects/SDL/src/stdlib/SDL_malloc.c:5306
> #5  0x00007ff704cde9ac in SDL_CreateMutex_REAL ()
>     at /home/maarten/projects/SDL/src/thread/pthread/SDL_sysmutex.c:34
> #6  0x00007ff704a353e4 in SDL_CreateProperties_REAL () at /home/maarten/projects/SDL/src/SDL_properties.c:159
> #7  0x00007ff704b1e739 in SDL_CreateThreadRuntime_REAL (fn=0x404c2e <SubThreadFunc>, name=0x7ff6c5fffd90 "Child8_5", 
>     userdata=0x7ff6c5fffde4, pfnBeginThread=0x0, pfnEndThread=0x0)
>     at /home/maarten/projects/SDL/src/thread/SDL_thread.c:399
> #8  0x00007ff704a64ccd in SDL_CreateThreadRuntime (a=0x404c2e <SubThreadFunc>, b=0x7ff6c5fffd90 "Child8_5", 
>     c=0x7ff6c5fffde4, d=0x0, e=0x0) at /home/maarten/projects/SDL/src/dynapi/SDL_dynapi_procs.h:127
> #9  0x0000000000404d06 in ThreadFunc (data=0x8) at /home/maarten/projects/SDL/test/torturethread.c:60
> #10 0x00007ff704b1e4de in SDL_RunThread (thread=0x13050740)
>     at /home/maarten/projects/SDL/src/thread/SDL_thread.c:323
> #11 0x00007ff704cde550 in RunThread (data=0x13050740)
>     at /home/maarten/projects/SDL/src/thread/pthread/SDL_systhread.c:69
> #12 0x00007ff7048ac897 in start_thread () from /lib64/libc.so.6
> #13 0x00007ff704933a5c in clone3 () from /lib64/libc.so.6
> ```

[This location](https://github.com/madebr/SDL/blob/9b9237a3b50c98c54f46c241c770bb188007da58/src/test/SDL_test_memory.c#L85). It looks like the memory tracker is not thread safe.

**Comment by slouken** [comment]
> [This location](https://github.com/madebr/SDL/blob/9b9237a3b50c98c54f46c241c770bb188007da58/src/test/SDL_test_memory.c#L85). It looks like the memory tracker is not thread safe.

That is very likely, and needs to be fixed if we're going to use it for threaded tests.

**Comment by madebr** [comment]
@icculus
Can you review my SDL_Mutex/SDL_SpinLock-free locking of the SDL_Test memory tracking?
I could not use those because they allocate memory internally.
It now uses a spinlock on top of `SDL_AtomicCompareAndSwap`.

File: src/test/SDL_test_memory.c [hunk 1]
```diff
@@ -67,6 +67,9 @@ static SDL_free_func SDL_free_orig = NULL;
 static int s_previous_allocations = 0;
 static SDL_tracked_allocation *s_tracked_allocations[256];
 static SDL_bool s_randfill_allocations = SDL_FALSE;
+static SDL_AtomicInt s_lock;
+#define LOCK_ALLOCATOR() do {} while (!SDL_AtomicCompareAndSwap(&s_lock, 0, 1))
```
Comment (slouken) [inline src/test/SDL_test_memory.c:71]: There's a definite possibility of thread starvation here. You should add SDL_CPUPauseInstruction() to the body of your loop.

**Comment by slouken** [comment]
> @icculus Can you review my SDL_Mutex/SDL_SpinLock-free locking of the SDL_Test memory tracking? I could not use those because they allocate memory internally. It now uses a spinlock on top of `SDL_AtomicCompareAndSwap`.

Aside from not pausing in your lock loop, this looks good.

**Review by slouken — APPROVED** [review]

**Comment by icculus** [comment]
>  Can you review my SDL_Mutex/SDL_SpinLock-free locking of the SDL_Test memory tracking?

Looks good to me.

**Pull request merged**